src/web3/useEthereumProvider.ts

```diff
@@ -0,0 +1,102 @@
+import { useEffect, useState } from 'react';
+import useScript from 'react-script-hook';
+
+import { createWeb3Plaid, PlaidFactory } from '../factory';
+import {
+  EthereumOnboardingOptions,
+  PlaidWeb3,
+  PlaidGlobalWithWeb3,
+} from '../types/web3';
+
+const PLAID_LINK_STABLE_URL =
+  'https://cdn.plaid.com/link/v2/stable/link-initialize.js';
+
+const noop = () => {};
+
+export const useEthereumProvider = (options: EthereumOnboardingOptions) => {
+  // Asynchronously load the plaid/link/stable url into the DOM
+  const [loading, error] = useScript({
+    src: PLAID_LINK_STABLE_URL,
+    checkForExisting: true,
+  });
+
+  // internal state
+  const [web3, setWeb3] = useState<PlaidWeb3 | null>(null);
+  const [plaid, setPlaid] = useState<PlaidFactory | null>(null);
+  const [iframeLoaded, setIframeLoaded] = useState(false);
+
+  useEffect(() => {
+    // If the link.js script is still loading, return prematurely
+    if (loading) {
+      return;
+    }
+
+    if (error || !window.Plaid) {
+      // eslint-disable-next-line no-console
+      console.error('Error loading Plaid', error);
+      return;
+    }
+
+    if (!web3) {
+      (window.Plaid as PlaidGlobalWithWeb3).web3().then(setWeb3);
+      return;
+    }
+
+    // If the token is undefined, return prematurely
+    if (!options.token) {
+      return;
+    }
+
+    // if an old plaid instance exists, destroy it before
+    // creating a new one
+    if (plaid != null) {
+      plaid.exit({ force: true }, () => plaid.destroy());
+    }
+
+    const next = createWeb3Plaid(
+      {
+        ...options,
+        onLoad: () => {
+          setIframeLoaded(true);
+          options.onLoad && options.onLoad();
+        },
+      },
+      web3.createEthereumOnboarding
+    );
+
+    setPlaid(next);
+
+    // destroy the Plaid iframe factory
+    return () => next.exit({ force: true }, () => next.destroy());
+  }, [loading, error, options.token, web3, setWeb3]);
+
+  const ready = plaid != null && (!loading || iframeLoaded) && web3 != null;
+
+  const openNoOp = () => {
+    if (!options.token) {
+      console.warn(
+        'react-plaid-link: You cannot call open() without a valid token supplied to useEthereumProvider. This is a no-op.'
+      );
+    }
+  };
+
+  const web3WithoutCreateEthereumOnboarding: Partial<PlaidWeb3> | null = ready
+    ? web3 && { ...web3 }
+    : null;
+  // Only expose the "static" methods
+  if (web3WithoutCreateEthereumOnboarding) {
+    delete web3WithoutCreateEthereumOnboarding['createEthereumOnboarding'];
+  }
```

Comment:
could you explain this? this seems a bit unconventional (sorry if i did not catch this in the initial internal review of this code)

Reply:
yeah so in addition to `.createEthereumOnboarding`, we expose other JS APIs to do other operations that don't strictly involve Link: https://plaid.com/docs/wallet-onboard/add-to-app/#reinitialize-previous-connections

So this bit 1) exposes those methods to developers via the hook 2) removes `createEthereumOnboarding` (which gets wrapped in `.open`) from both the types and actual returned object

I did this implementation so that anything we attach to `PlaidWeb3` in the future will automatically cascade out of the hook (though it will lack types, in retrospect) - but we could also manually expose them and that would get rid of the `delete` business.

wydt? Reading through it again, it might be cleaner to just manually export each method anyway since they would lack types otherwise / require updating the dependency anyway.

Reply:
updated it - it was also causing some kind of typescript compile issue inside storybook

Reply:
Agreed, I don't think we should automatically expose methods that don't have types. Thanks for addressing